_posts/2019-03-09-rest-response-size.md

```diff
@@ -0,0 +1,138 @@
+
+## Sending too much data
+
+Thoughtfully developing an API endpoint requires some thought into what data is or is not being sent over the wire. While it takes little effort to add a new field to your response object, it does take some thought into _if_ the field should be added in the first place. One of the best ways to mitigate this is by embracing Hypermedia with your APIS. Lets use an example of a users response object before we add Hypermedia. 
+
+```json
+{
+  "data": [
+    {
+      "type": "User",
+      "id": 1,
+      "attributes": {
+        "userId": "11122",
+        "name": "Crashy McCrashface",
+        "position": "Lead Bike Engineer",
+        "yearsAtPosition": 2,
+        "previousPosition": "Bike Engineer",
+        "managerId": "10109"
+      }
+    },
+    {
+      "type": "User",
+      "id": 2,
+      "attributes": {
+        "userId": "11123",
+        "name": "Turtle McTurtleface",
+        "position": "Lead Swamp Designer",
+        "yearsAtPosition": 5,
+        "previousPosition": "Junior Swamp Designer",
+        "managerId": "10002"
+      }
+    }
+  ];
+}
+```
+
+Looking at the data above, there is so much included that doesnt need to be in a ```HTTP/2 GET /users``` request. Lets simplify the whole thing. 
+
+```json
+{
+  "data": [
+    {
+      "type": "User",
+      "id": 1,
+      "attributes": {
+        "userId": "11122",
+        "name": "Crashy McCrashface"
+      },
+      "relationships": {
+        "user": {
+          "links": {
+            "self": "https://apisyouwonthate.com/api/users/11122/relationships/user",
+            "related": "https://apisyouwonthate.com/api/users/11122",
+          },
+          "data": {
+            "type": "User",
+            "id": "11122",
+          }
+        } 
+      }
+    },
+    {
+      "type": "User",
+      "id": 2,
+      "attributes": {
+        "userId": "11123",
+        "name": "Turtle McTurtleface"
+      },
+      "relationships": {
+        "user": {
+          "links": {
+            "self": "https://apisyouwonthate.com/api/users/11123/relationships/user",
+            "related": "https://apisyouwonthate.com/api/users/11123",
+          },
+          "data": {
+            "type": "User",
+            "id": "11123",
+          }
+        } 
+      }
+    }
+  ];
+}
+```
+
+While this has more structure to the request, the data being sent back has been cut down and we give the client user instructions to where to find the data they may want next. This is called "discoverability". Now, if we request ```HTTP/2 GET /users/11123``` we would want to return something similar to the first example, like so: 
+
+```json
+{
+  "data": [
+    {
+      "type": "User",
+      "id": 1,
+      "attributes": {
```

Comment:
If you keep this example then maybe split it out into "positions" or "roles" and explain that the extra split is actually a useful feature: you can have multiple roles at the company, you may switch roles, slowly transitin from one to the other with overlap, etc. Usually the extra resources provide extra functionality. :)

Reply:
Im going to go with the invoice example, seems more straight forward :)